daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala

```diff
@@ -978,17 +978,12 @@ final class VariableRuntimeData(
   @throws(classOf[java.io.IOException])
   final private def writeObject(out: java.io.ObjectOutputStream): Unit = serializeObject(out)
 
-  private lazy val state =
-    if (!maybeDefaultValueExpr.isDefined) VariableUndefined
-    else VariableDefined
-
-  private lazy val value: DataValuePrimitiveNullable =
-    if (maybeDefaultValueExpr.isEmpty) DataValue.NoValue
-    else {
-      val defaultValueExpr = maybeDefaultValueExpr.get
-      defaultValueExpr match {
-        case constExpr: ConstantExpression[_] => DataValue.unsafeFromAnyRef(constExpr.constant)
-        case _ => DataValue.NoValue
+  private lazy val (state, value) = {
+    if (!maybeDefaultValueExpr.isDefined)
+      (VariableUndefined, DataValue.NoValue)
+    else maybeDefaultValueExpr.get match {
+        case constExpr: ConstantExpression[_] => (VariableDefined, DataValue.unsafeFromAnyRef(constExpr.constant))
+        case _ => (VariableUndefined, DataValue.NoValue)
```

Comment:
Ah, so this is how you handle constants? I'm curious what others say, but I'd rather not have this special case here. It's maybe a little more performant when it is a constant, but it adds extra special cases and spreads out the NVI default logic in separate places. And I suspect NVI default values will rarely be constants, so we probably won't even hit this case very often.

I'd rather we just delete this state/value thing entirely, change the createVariableInstance function always creates a new instances with VariableUndefinedState and NoValue, and then the parsers/unparsers can mutate the state in a single place.

Reply:
That makes sense to me.  Most of the stuff in RuntimeData was set up to just get things working with the way we had been handling variables before when they were much simpler things without so much to keep track of.

Reply:
Also, just realized that removing the state/value stuff from RuntimeData.scala means we don't need to do the complicated/weird lazy argument stuff.

Reply:
One issue with this approach is that it does break some tests that relied on default values with constant expressions being evaluated at compile time (a few of the external variable loader unit tests (non-tdml).  I'm going to push up a commit with them changed to not use default values.

Reply:
Actually, looking at the tests that are failing, they are all external variable tests that don't use tdml runner.  I'm wondering if these are simply relics of a time when TDMLRunner didn't support external variables.  If so, could we just remove them (after assuring that there are equivalent TDML tests)?

Reply:
Doesn't seem unreasonable to me :+1: